packages/block-editor/src/hooks/block-bindings.js

```diff
@@ -66,57 +67,117 @@ function BlockBindingsPanelMenuContent( { fieldsList, attribute, binding } ) {
 		},
 		[ clientId, attribute ]
 	);
+
 	return (
-		<>
-			{ Object.entries( fieldsList ).map( ( [ name, fields ], i ) => (
-				<Fragment key={ name }>
-					<Menu.Group>
-						{ Object.keys( fieldsList ).length > 1 && (
-							<Menu.GroupLabel>
-								{ registeredSources[ name ].label }
-							</Menu.GroupLabel>
-						) }
-						{ Object.entries( fields )
-							.filter(
-								( [ , args ] ) => args?.type === attributeType
-							)
-							.map( ( [ key, args ] ) => (
-								<Menu.RadioItem
-									key={ key }
-									onChange={ () =>
-										updateBlockBindings( {
-											[ attribute ]: {
-												source: name,
-												args: { key },
-											},
-										} )
-									}
-									name={ attribute + '-binding' }
-									value={ key }
-									checked={ key === currentKey }
-								>
-									<Menu.ItemLabel>
-										{ args?.label }
-									</Menu.ItemLabel>
-									<Menu.ItemHelpText>
-										{ args?.value }
-									</Menu.ItemHelpText>
-								</Menu.RadioItem>
-							) ) }
-					</Menu.Group>
-					{ i !== Object.keys( fieldsList ).length - 1 && (
-						<Menu.Separator />
-					) }
-				</Fragment>
-			) ) }
-		</>
+		<Menu placement={ isMobile ? 'bottom-start' : 'left-start' }>
+			{ Object.entries( sources ).map( ( [ sourceKey, source ] ) => {
+				// Only show sources that have compatible data for this specific attribute.
+				const sourceDataItems = source.data?.filter(
+					( item ) => item?.type === attributeType
+				);
+
+				// Skip sources with no compatible data
+				if ( ! sourceDataItems || sourceDataItems.length === 0 ) {
+					return null;
+				}
+
+				if ( source.mode === 'dropdown' ) {
+					return (
+						<Menu
+							key={ sourceKey }
+							placement={
+								isMobile ? 'bottom-start' : 'left-start'
+							}
+						>
+							<Menu.SubmenuTriggerItem>
+								<Menu.ItemLabel>
+									{ source.label }
+								</Menu.ItemLabel>
+							</Menu.SubmenuTriggerItem>
+							<Menu.Popover gutter={ 8 }>
+								<Menu.Group>
+									{ sourceDataItems.map( ( item ) => (
+										<Menu.CheckboxItem
+											key={ item.key }
```

Comment:
If `item.key` is truly optional, the `key` prop should have a fallback value. Otherwise I get the expected React warning `Each child in a list should have a unique "key" prop.` Maybe `source.name` plus array index?

```suggestion
									{ sourceDataItems.map( ( item, index ) => (
										<Menu.CheckboxItem
											key={ item.key ?? `${ source.name }/${ index }`
```

Reply:
Taking into account that the list of items for the source depends on the context, it would be better to avoid using `index` in the key, as it might change between re-renders, leading to some unexpected behaviors (see [React docs](https://react.dev/learn/rendering-lists#why-does-react-need-keys)). However, I'm not sure if there is a simple way to create a unique key based on the `item`, object.

> You might be tempted to use an item’s index in the array as its key. In fact, that’s what React will use if you don’t specify a key at all. But the order in which you render items will change over time if an item is inserted, deleted, or if the array gets reordered. Index as a key often leads to subtle and confusing bugs.